Rewrite description array.prototype.reduce (#12949)
mdn/content: files/en-us/web/javascript/reference/global_objects/array/reduce/index.md

If an initial value is provided, `reduce()` calls the "reducer" callback function on each element in the array, in order. If no initial value is provided, `reduce()` calls the callback function on each element in the array after the first element.

`reduce` returns whatever is returned from the callback on the final iteration.

### When to not use reduce()

Recursive functions like `reduce()` can be powerful but sometimes difficult to understand,
especially for less experienced JavaScript developers.
If code becomes clearer when using other array methods,
developers must weigh the readability tradeoff against the other benefits
of using `reduce()`.
In cases where `reduce()` is the best choice, documentation and semantic variable
naming can help mitigate readability drawbacks.

### Behavior during array mutations

`reduce` does not mutate the array it is used on, however it is possible for code inside the callback to mutate
the array. Here is what happens if a mutation occurs:

- If elements are appended to the array after `reduce()` begins to iterate over the array, the callback
function does not iterate over the appended elements.

> - If existing elements of the array are changed, their values as passed to the callback function will be the values at the time `reduce()` iterates over the elements.

Author: @dipikabh let's have a look back at the spec and lean on that language here:

> If existing elements of the array are changed, their value as passed to callbackfn will be the value at the time reduce visits them

I see now that the prose above does not quite match the preciseness of this spec statement, mainly around the use of the word "iterate." Maybe we should put this back to be as close to the spec as we can, with teeny readability changes:

> If existing elements of the array do get changed, the values passed to the callback will be the values from the time that reduce was first called on the array.

How does that sound?

> - If existing elements of the array are changed, their values as passed to the callback function will be the values at the time `reduce()` iterates over the elements.

Reviewer: That is definitely much better! Thanks for revisiting the spec and rewriting.
Just few minor updates:

> If existing elements of the array do get changed, the values passed to the callback function will be the values from the time that `reduce()` was first called on the array.

> - If existing elements of the array are changed, their values as passed to the callback function will be the values at the time `reduce()` iterates over the elements.

Author: Updated!

## Description

`reduce` takes two arguments, a callback function and an optional initial value.

> `reduce` calls the callback, as a function, once for each element in the array whenever
> an initial value is provided, in ascending order. If no initial value is provided,
> `reduce` calls the callback once for each element in the array after the first element.

Reviewer: ```suggestion
If an initial value is provided, `reduce()` calls the "reducer" callback function on each element in the array, in order. If no initial value is provided, `reduce()` calls the callback function on each element in the array after the first element.
```
- Reworded the first sentence so that both sentences follow the same flow.
- Replaced "callback" with ""reducer" callback function" for consistency with the terminology used before this section
- The spec does say "ascending order". But as in the first paragraph on this page, saying "in order" will perhaps be less confusing (?).

`reduce` returns whatever is returned from the callback on the final iteration.

### When to not use reduce()

> Recursive functions like `reduce` can be powerful, but sometimes difficult to understand,
> especially for less experienced JavaScript developers.
> If code becomes clearer when using other array methods,
> developers must weigh the readability tradeoff against the other benefits
> of using reduce.
> In cases where reduce is the best choice, documentation and semantic variable
> naming can help mitigate readability drawnbacks.

Reviewer: ```suggestion
Recursive functions like `reduce()` can be powerful but sometimes difficult to understand,
especially for less experienced JavaScript developers.
If code becomes clearer when using other array methods,
developers must weigh the readability tradeoff against the other benefits
of using `reduce()`.
In cases where `reduce()` is the best choice, documentation and semantic variable
naming can help mitigate readability drawbacks.
```
- Removed the comma before "but" because here one of the two clauses being joined by "but" is not a complete sentence

> 
> Recursive functions like `reduce` can be powerful, but sometimes difficult to understand,
> especially for less experienced JavaScript developers.
> If code becomes clearer when using other array methods,
> developers must weigh the readability tradeoff against the other benefits
> of using reduce.
> In cases where reduce is the best choice, documentation and semantic variable
> naming can help mitigate readability drawnbacks.
> 

Reviewer 2: ```suggestion

Recursive functions like `reduce` can be powerful, but sometimes difficult to understand,
especially for less-experienced JavaScript developers.
If code becomes more clear when using other array methods,
developers must weigh the readability tradeoff against the other benefits
of using `reduce()`.
In cases where `reduce()` is the best choice, documentation and well-chosen variable
names can help mitigate readability drawbacks.

```

### Behavior during array mutations

`reduce` does not mutate the array it is used on, however it is possible for code inside the callback to mutate
the array.

> If elements are appended to the array after `reduce` begins to iterate over the array, the appended
> elememts are not visited by *callbackfn*.

Reviewer: ```suggestion
If elements are appended to the array after `reduce()` begins to iterate over the array, the appended
elements are not iterated by the callback function.
```
- Another rewrite suggestion to change the sentence to active voice:
"If elements are appended to the array after `reduce()` begins to iterate over the array, the callback function does not iterate over the appended elements."

Author: I applied the second suggestion, thanks!

> If existing elements of the array are changed, their value as passed to *callbackfn* will be the value at the time `reduce` visits them.

Reviewer: ```suggestion
If existing elements of the array are changed, their values as passed to the callback function will be the values at the time `reduce()` iterates over the elements.
```

> Elements that are deleted after the call to `reduce` begins and before being visited are not visited.

Reviewer: ```suggestion
Elements that are deleted after the call to `reduce()` begins and before being iterated over are not visited.
```

Author: I think I want to take another pass in a followup PR to change some of this preexisting verbage. This is a pretty confusing sentence!

Author: I made this a bulleted list now.
